### PR TITLE
Optimize `Replacing` merge algorithm for non-intersecting parts

### PR DESCRIPTION
**Pull request by CurtizJ**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Optimized `Replacing` merge algorithm for non intersecting parts.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit b4aad3dd590d82428f41d0c425c719a3c9a7d30e with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70977/b4aad3dd590d82428f41d0c425c719a3c9a7d30e/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/ac766ae2d81564ceadc717122952ada2c964cd20/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp [hunk 1]
```diff
@@ -113,30 +135,73 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
...
-                }
-                else
-                    insertRow();
-                /// insertRow() may has not been called
-                saveChunkForSkippingFinalFromSelectedRow();
+                auto replace_final_selection = ColumnUInt64::create(chunk_num_rows);
+                auto & replace_final_data = replace_final_selection->getData();
+
+                std::iota(replace_final_data.begin(), replace_final_data.end(), 0);
+                current_chunk.getChunkInfos().add(std::make_shared<ChunkSelectFinalIndices>(std::move(replace_final_selection)));
```
Comment (nickitat) [inline src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp:172]: i understand that it is the same in current master, but looks wasteful. could we handle this case separately w/o creating a column explicitly?

File: src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp [hunk 2]
```diff
@@ -113,30 +135,73 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
...
+
+                Status status(std::move(current_chunk), false);
```
Comment (nickitat) [inline src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp:174]:
```suggestion
                Status status(std::move(current_chunk));
```

File: src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp [hunk 3]
```diff
@@ -113,30 +135,73 @@ IMergingAlgorithm::Status ReplacingSortedAlgorithm::merge()
...
+            /// Write order of rows for other columns this data will be used in gather stream
+            if (out_row_sources_buf)
+            {
+                /// All rows are not skipped.
+                RowSourcePart row_source(source_num);
+                for (size_t i = 0; i < chunk_num_rows; ++i)
+                    out_row_sources_buf->write(row_source.data);
+            }
+
+            Status status(merged_data->pull(), false);
```
Comment (nickitat) [inline src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp:191]:
```suggestion
            Status status(merged_data->pull());
```

File: src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp (hunk 1 above)
Comment (CurtizJ) [reply src/Processors/Merges/Algorithms/ReplacingSortedAlgorithm.cpp]: Yes, sure.

**Review by nickitat — APPROVED** [review]

**Pull request merged**